boot/zephyr/Kconfig

```diff
@@ -346,4 +346,12 @@ config UPDATEABLE_IMAGE_NUMBER
 	help
 	  Enables support of multi image update.
 
+config MCUBOOT_DOWNGRADE_PREVENTION
+	bool "Downgrade prevention"
+	depends on BOOT_UPGRADE_ONLY
```

Comment:
Can you explain further why this depends on upgrade only? We can check before starting an upgrade. There are plenty of other issues with a malicious user manipulating swap, but that would be caught at the end, and even in that case, trying to prevent an upgrade is at least better than not doing it at all. Any thoughts on how this could be used with swap?

To get this in, we should have some simulation tests for this functionality.

We should also think about naming here, to make sure it doesn't end up too confusing. There is another feature that will be coming from TFM that adds support for a hardware anti-rollback counter. That will protect against even an attacker that can replace the primary image. But, not all devices have this counter, so we definitely want this software version.

Reply:
I assume the following (please let me know if any of these are incorrect):
1. The only reason to perform a swap instead of an overwrite is to allow rollback.
2. Whether or not the rollback is performed depends solely on data found in the image trailers, assuming that the image hashes are correct.
3. The image trailers are not protected from modification in any way. (*)
4. Given 1, 2 and 3 it is clear that the application booted by MCUboot can trigger a swap operation by modifying the contents of the image trailers.
5. Given 4, it is not possible for MCUBoot to distinguish between if the image trailers has been written by MCUBoot or a malicious application.

Because of 1, MCUBoot with downgrade prevention has to allow a "REVERT" operation to be performed with a lower version number (otherwise there is no point in supporting swapping).
Because of 4 and 5 it is possible to break the downgrade prevention by modifying the image trailers.

(*)
If primary and secondary is open, simply write 0xff to image OK to bypass downgrade prevention
If primary is locked and secondary is open during first boot -> downgrade prevention can be bypassed by downloading an old image to secondary, and not writing IMAGE-OK, which will trigger the old image to be swapped in.
The solution to this problem is to lock the secondary slot during TEST boot, and then lock primary slot in subsequent boots.
Another solution to this problem is (as you point out) monotonic version counters (anti rollback counter).

Regarding naming confusion, these two features provide the same functionality, so one would typically choose one of them.
Since the term "rollback" is used in MCUBoot to describe the two step upgrade process, I prefer to call this feature "downgrade prevention" in this context.

In the future we could have an option to select between the two variants (SW or HW based).

Reply:
Thanks for elaboration. Looks like downgrade prevention witch swap make sense only if we can trust to image trailers. Other actors here are application images. May for some reason/some case the image be trusted (as it has signature verified)?

Reply:
> Looks like downgrade prevention witch swap make sense only if we can trust to image trailers.

Note that downgrade prevention with swapping is possible if we implement the scheme proposed above, where you:
1. During TEST boot of newly swapped image, lock the secondary slot.
2. During second boot after image has self-verified (IMAGE-OK), lock the primary slot

>  Other actors here are application images. May for some reason/some case the image be trusted (as it has signature verified)?

No, you can not trust the image. 

Firstly there is no guarantee that the application has any level of flash-write protection. So bugs in the application could allow modifications in run time, and these changes would not be discovered until after reboot, which is too late.

Secondly, some attacks do not modify the firmware, such as [return oriented programming](https://en.wikipedia.org/wiki/Return-oriented_programming)

Reply:
> Note that adding flash protection support of the primary and secondary slot as described in a different comment can give downgrade prevention with swapping.

That's good. Such protection mechanism sounds like a another patch objective.

Reply:
This catches the misconfiguration in Zephyr, but as I just discovered, the code seems perfectly happy to build without overwrite only, but with downgrade protection enabled. I think this configuration should cause a compilation error, rather than just failing silently.